packages/client/src/object/formatting/formatNumber.ts

```
@@ -0,0 +1,398 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type {
+  NumberFormatAffix,
+  NumberFormatCurrency,
+  NumberFormatCustomUnit,
+  NumberFormatOptions,
+  NumberFormatRatio,
+  NumberFormatScale,
+  NumberFormatStandardUnit,
+  NumberRatioType,
+  NumberScaleType,
+  PropertyNumberFormattingRuleType,
+} from "@osdk/api";
+import type { SimpleOsdkProperties } from "../SimpleOsdkProperties.js";
+import { resolvePropertyReference } from "./propertyFormattingUtils.js";
+
+/**
+ * Extended NumberFormatOptions that includes rounding mode support
+ * which is available in modern browsers but not yet in TypeScript types.
+ * Only includes modes that are actually supported by Intl.NumberFormat and used in our API.
+ */
+export interface ExtendedNumberFormatOptions extends Intl.NumberFormatOptions {
+  roundingMode?: "ceil" | "floor" | "halfExpand";
+  convertNegativeToParenthesis?: boolean;
+}
+
+/**
+ * Formats a number value according to the specified formatting rule
+ */
+export function formatNumber(
+  value: number,
+  numberType: PropertyNumberFormattingRuleType,
+  objectData: SimpleOsdkProperties,
+  locale: string,
+): string | undefined {
+  switch (numberType.type) {
+    case "standard":
+      return formatStandardNumber(
+        value,
+        numberType.baseFormatOptions,
+        locale,
+      );
+
+    case "fixedValues":
+      return numberType.values[value] ?? value.toString();
```

Should we throw here/return undefined?